File: apps/server/src/config/billing-mode.ts
export type BillingMode = "hosted" | "self_hosted";

function resolveBillingMode(): BillingMode {
  const configured = process.env.QUICKVOICE_BILLING_MODE?.trim().toLowerCase();
  if (configured === "hosted") return "hosted";
  if (configured === "self_hosted" || configured === "self-hosted") {
    return "self_hosted";
  }
  if (configured) {
    throw new Error(
      "QUICKVOICE_BILLING_MODE must be either 'hosted' or 'self_hosted'",
    );
  }

  // Production defaults to metered billing so a missing flag cannot silently
  // make the hosted service free. Local/community development remains
  // unmetered unless hosted mode is selected explicitly.
  return process.env.NODE_ENV === "production" ? "hosted" : "self_hosted";
}

export const billingMode = resolveBillingMode();
export const isHostedBilling = billingMode === "hosted";

File: apps/ai/utils/billing_mode.py
from __future__ import annotations

import os
from collections.abc import Mapping


def resolve_billing_mode(environ: Mapping[str, str] | None = None) -> str:
    env = os.environ if environ is None else environ
    configured = str(env.get("QUICKVOICE_BILLING_MODE", "")).strip().lower()
    if configured == "hosted":
        return "hosted"
    if configured in {"self_hosted", "self-hosted"}:
        return "self_hosted"
    if configured:
        raise ValueError(
            "QUICKVOICE_BILLING_MODE must be either 'hosted' or 'self_hosted'"
        )
    return (
        "hosted"
        if str(env.get("NODE_ENV", "")).strip().lower() == "production"
        else "self_hosted"
    )


def hosted_billing_required(environ: Mapping[str, str] | None = None) -> bool:
    return resolve_billing_mode(environ) == "hosted"
